engine/servermaster/jobop/backoff.go

```diff
@@ -0,0 +1,194 @@
+func WithMultiplier(multiplier float64) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.multiplier = multiplier
+	}
+}
+
+// WithClocker sets clocker to backoffOpts
+func WithClocker(clocker clock.Clock) BackoffOption {
+	return func(opts *backoffOpts) {
+		opts.clocker = clocker
+	}
+}
+
+type backoffEvent struct {
+	tp backoffEventType
+	ts time.Time
+}
+
+// NewJobBackoff creates a new job backoff
+func NewJobBackoff(jobID string, options ...BackoffOption) *JobBackoff {
+	opts := &backoffOpts{
+		resetInterval:   defaultBackoffRestInterval,
+		initialInterval: defaultBackoffInitInterval,
+		maxInterval:     defaultBackoffMaxInterval,
+		multiplier:      defaultBackoffMultiplier,
+		clocker:         clock.New(),
+	}
+	for _, option := range options {
+		option(opts)
+	}
+
+	errBackoff := backoff.NewExponentialBackOff()
+	errBackoff.InitialInterval = opts.initialInterval
+	errBackoff.MaxInterval = opts.maxInterval
+	errBackoff.Multiplier = opts.multiplier
+	errBackoff.Reset()
+
+	return &JobBackoff{
+		jobID:      jobID,
+		opts:       opts,
+		errBackoff: errBackoff,
+	}
+}
+
+// JobBackoff is a job backoff manager, it recoreds job online and offline events
+// and determines whether a job can be re-created based on backoff mechanism.
+// The backoff stragegy is as following
+// - Each time a fail event arrives, the backoff time will be move forward by
+//   nextBackoff.
+// - If a job is success for more than `resetInterval`, the backoff history will
+//   be cleared, and backoff time will be re-calculated.
+type JobBackoff struct {
+	jobID           string
+	events          []backoffEvent
+	opts            *backoffOpts
+	errBackoff      *backoff.ExponentialBackOff
+	backoffInterval time.Duration
+}
+
+// Allow returns whether new request(create job) is allowd
+func (b *JobBackoff) Allow() bool {
+	var lastErrorTime time.Time
+	for i := len(b.events) - 1; i >= 0; i-- {
+		event := b.events[i]
+		if event.tp == backoffOffline {
+			lastErrorTime = event.ts
+			break
+		}
+	}
+	return b.opts.clocker.Since(lastErrorTime) >= b.backoffInterval
+}
+
+// Success is called when a success event happens
+func (b *JobBackoff) Success() {
+	event := backoffEvent{
+		tp: backoffOnline,
+		ts: b.opts.clocker.Now(),
+	}
+	b.addEvent(event)
+}
+
+// Fail is called when a failure event happens
+func (b *JobBackoff) Fail() {
+	event := backoffEvent{
+		tp: backoffOffline,
+		ts: b.opts.clocker.Now(),
+	}
+	b.addEvent(event)
+	b.nextBackoff()
+}
+
+// addEvent appends new backoff event into backoffer
+func (b *JobBackoff) addEvent(event backoffEvent) {
+	// The last event is online and it is earlier than `resetInterval`,
+	// reset the backoff
+	if len(b.events) > 0 {
+		lastEvent := b.events[len(b.events)-1]
+		if lastEvent.tp == backoffOnline &&
+			b.opts.clocker.Since(lastEvent.ts) >= b.opts.resetInterval {
+			b.events = make([]backoffEvent, 0)
+			b.resetErrBackoff()
+		}
+	}
+	b.events = append(b.events, event)
```

Comment:
Actually, we only need `lastOnlineTs` to do this check.
```
 if lastOnlineTs >0 && b.opts.clocker.Since(lastEvent.ts) >= b.opts.resetInterval {
			b.resetErrBackoff()
}
```

Set lastOnlineTs = 0 every time we receive the `offline`  event. 
Set lastOnlineTs = xxx every time we receive the `online`  event.